backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/loops/ProgressionHandlers.kt

```diff
@@ -0,0 +1,148 @@
+import org.jetbrains.kotlin.ir.types.*
+import org.jetbrains.kotlin.ir.visitors.IrElementVisitor
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+
+internal class RangeToHandler(val progressionElementClasses: Collection<IrClassSymbol>) : ProgressionHandler {
+    override val matcher = SimpleCalleeMatcher {
+        dispatchReceiverRestriction { it != null && it.type.classifierOrNull in progressionElementClasses }
+        fqNameRestriction { it.pathSegments().last() == Name.identifier("rangeTo") }
+        parameterCountRestriction { it == 1 }
+        parameterRestriction(0) { it.type.classifierOrNull in progressionElementClasses }
+    }
+
+    override fun build(call: IrCall, progressionType: ProgressionType) =
+            ProgressionInfo(progressionType, call.dispatchReceiver!!, call.getValueArgument(0)!!)
+}
+
+internal class DownToHandler(val progressionElementClasses: Collection<IrClassSymbol>) : ProgressionHandler {
+    override val matcher = SimpleCalleeMatcher {
+        singleArgumentExtension(FqName("kotlin.ranges.downTo"), progressionElementClasses)
+        parameterRestriction(0) { it.type.classifierOrNull in progressionElementClasses }
+    }
+
+    override fun build(call: IrCall, progressionType: ProgressionType): ProgressionInfo? =
+            ProgressionInfo(progressionType,
+                    call.extensionReceiver!!,
+                    call.getValueArgument(0)!!,
+                    increasing = false)
+}
+
+internal class UntilHandler(val progressionElementClasses: Collection<IrClassSymbol>) : ProgressionHandler {
+    override val matcher = SimpleCalleeMatcher {
+        singleArgumentExtension(FqName("kotlin.ranges.until"), progressionElementClasses)
+        parameterRestriction(0) { it.type.classifierOrNull in progressionElementClasses }
+    }
+
+    override fun build(call: IrCall, progressionType: ProgressionType): ProgressionInfo? =
+            ProgressionInfo(progressionType,
+                    call.extensionReceiver!!,
+                    call.getValueArgument(0)!!,
+                    closed = false)
+}
+
+internal class IndicesHandler(val context: Context) : ProgressionHandler {
+
+    private val symbols = context.ir.symbols
+
+    private val supportedArrays = symbols.primitiveArrays.values + symbols.array
+
+    override val matcher = SimpleCalleeMatcher {
+        extensionReceiverRestriction { it != null && it.type.classifierOrNull in supportedArrays }
+        fqNameRestriction { it == FqName("kotlin.collections.<get-indices>") }
+        parameterCountRestriction { it == 0 }
+    }
+
+    override fun build(call: IrCall, progressionType: ProgressionType): ProgressionInfo? {
+        val int0 = IrConstImpl.int(call.startOffset, call.endOffset, context.irBuiltIns.intType, 0)
+
+        val bound = with(context.createIrBuilder(call.symbol, call.startOffset, call.endOffset)) {
+            val clazz = call.extensionReceiver!!.type.classifierOrFail
```

Comment:
Btw, this is incorrect. Consider something like
```
fun <T : ByteArray> T.foo() {
     for (i in this.indices) {}
}
```

Reply:
Added restriction on the actual receiver's type.

Reply:
Actually it would be enough to check function for extension receiver *parameter* type.